File: src/lexflow/api/routers/departments.py
"""``GET /api/v1/departments`` — the issuing-department (ministerio)
vocabulary across the corpus (#671 gap B).

Aggregates each law's ``department`` frontmatter field (BOE metadata, parsed
straight through with no normalisation) into a count-ranked vocabulary. The
frontend uses it for the Explorer department facet.

--- WHERE TO CHANGE IF X CHANGES ---
* Department extraction → ``lexflow.core.parser.frontmatter_to_metadata``.
* Aggregation            → ``LawRegistry.department_counts``.
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends

from lexflow.api.dependencies import get_law_registry
from lexflow.core.registry import LawRegistry
from lexflow.core.schemas import DepartmentCount, DepartmentsResponse

router = APIRouter(tags=["Departments"])


@router.get(
    "/departments",
    response_model=DepartmentsResponse,
    summary="Issuing department (ministerio) vocabulary, ranked by usage.",
)
def list_departments(
    registry: Annotated[LawRegistry, Depends(get_law_registry)],
) -> DepartmentsResponse:
    """Return ``{items: [{department, count}]}`` sorted by count desc, then name asc."""
    items = [
        DepartmentCount(department=department, count=count) for department, count in registry.department_counts()
    ]
    return DepartmentsResponse(items=items)

File: frontend/src/lib/api/departments.ts
/**
 * `liveApi.departments` — `[{department, count}]` vocabulary ranked by
 * corpus usage (#671 gap B).
 *
 * Empty until the corpus frontmatter carries a `department` field; the
 * endpoint + shape are live so the Explorer department filter works the
 * moment any law is attributed to a ministerio.
 */

import type { ApiClient, DepartmentCount } from '../types';
import { http } from './http';

export const liveDepartmentsApi: ApiClient['departments'] = {
  list: async () => {
    // Sprint 6 api-6 wrapper convention — `{items: [...]}`, mirrors `tags.ts`.
    const raw = await http<{ items: DepartmentCount[] }>('/departments');
    return raw.items;
  },
};
